### ISSUE TITLE
PageInfo type field nullability different from specification

### ISSUE THREAD

**Issue by Cito** [graphql/graphql-relay-js#243]
The [relay spec says](https://relay.dev/graphql/connections.htm#sec-undefined.PageInfo.Fields) that "`PageInfo` must contain fields `hasPreviousPage` and `hasNextPage`, both of which return _non‐null_ booleans. It must also contain fields `startCursor` and `endCursor`, both of which return _non‐null_ opaque strings."

However, the [Flow type for PageInfo](https://github.com/graphql/graphql-relay-js/blob/ecc478e7f6ed74c233aa53bbf5e9d4e04dffeada/src/connection/connectiontypes.js#L18) makes all of them nullable, and the [GraphQL object type for PageInfo](https://github.com/graphql/graphql-relay-js/blob/ecc478e7f6ed74c233aa53bbf5e9d4e04dffeada/src/connection/connection.js#L135) makes `startCursor` and `endCursor` nullable.

Can we fix these types to conform to the specification?

**Comment by Cito** [comment]
This should also solve #240.

**Comment by olso** [comment]
Just ran into this

![Screenshot 2020-07-16 at 17 58 14](https://user-images.githubusercontent.com/7975158/87694069-00bea900-c78e-11ea-9e64-1bcf990d9607.png)

**Comment by numberten** [comment]
Even though the spec claims `startCursor` and `endCursor` must be non-null, I believe they need to support null to handle where the edges return an empty list.

**Comment by GabrielGil** [comment]
Indeed. I have run into the same incongruence, and it is not clear what option is correct. Either the spec is just ambiguous or out of date, or maybe this library is just taking a different approach.

Either way, in case the spec is respected, and both `startCursor` and `endCursor` are _not null opaque strings_, what should be their value when there are no results? Should the connection be null? Should the cursors be an empty string?

**Comment by Cito** [comment]
It seems this has been (accidentally?) solved by #300 already.

**Comment by bamorim** [comment]
@Cito thanks for reporting that, just stumbled upon this as I was confused about the spec and some implementations out in the wild.

I seems that #300 still marks the `endCursor` and `startCursor` as nullables:

```js
  startCursor: ConnectionCursor | null,
  endCursor: ConnectionCursor | null,
```

https://github.com/graphql/graphql-relay-js/pull/300/files#diff-c906c9bb7d4a80f721e5ef3e8c7159dda11e5f8829ace2f04982a460a453c0a2R12-R13

It only fixes the nullability of `hasPreviousPage` and `hasNextPage`.

**Comment by IvanGoncharov** [comment]
@Cito Thanks for reporting it spec should definitely be clarified about the values of `startCursor` and `endCursor` in case of an empty array.
